### ISSUE TITLE
unable to sync from freezer - parlia consensus failure

### ISSUE THREAD

**Issue by immibis** [bnb-chain/bsc#1769]
#### System information

Geth
Version: 1.2.8
Git Commit: (v1.2.8 fd58e6f0ac40d443bf4b4aed68740c4fc0501a34 with additional logging added)
Architecture: amd64
Go Version: go1.20.5
Operating System: linux (kernel 5.15.41)

#### Expected behaviour

geth syncs from freezer

#### Actual behaviour

```
INFO [07-15|02:32:38.014|geth/main.go:297] Starting Geth on Ethereum mainnet... 
INFO [07-15|02:32:38.017|cmd/utils/flags.go:1308] Maximum peer count                       ETH=100 LES=0 total=100
INFO [07-15|02:32:38.017|cmd/utils/flags.go:1405] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
WARN [07-15|02:32:38.017|cmd/utils/flags.go:1373] Option nousb is deprecated and USB is deactivated by default. Use --usb to enable 
INFO [07-15|02:32:38.018|cmd/utils/flags.go:1791] Set global gas cap                       cap=50,000,000
INFO [07-15|02:32:38.018|eth/backend.go:140]      Allocated trie memory caches             clean=0.00B dirty=0.00B
INFO [07-15|02:32:38.018|ethdb/leveldb/leveldb.go:117] Allocated cache and file handles         database=<path redacted>/geth_data/geth/chaindata cache=16.00MiB handles=2048
INFO [07-15|02:32:38.071|core/rawdb/freezer.go:376]    AncientDB item count                     items=26,856,810
INFO [07-15|02:32:38.071|core/rawdb/freezer.go:173]    Opened ancient database                  database=<path redacted>/ancient readonly=false frozen=26,856,810
INFO [07-15|02:32:38.072|eth/backend.go:156]           Initialised chain configuration          config="{ChainID: 56 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0, Muir Glacier: 0, Ramanujan: 0, Niels: 0, MirrorSync: 5184000, Bruno: 13082000, Berlin: <nil>, YOLO v3: <nil>, CatalystBlock: <nil>, London: <nil>, ArrowGlacier: <nil>, MergeFork:<nil>, Euler: 18907621, Gibbs: 23846001, Nano: 21962149, Moran: 22107423, Planck: 27281024,Luban: 29020050, Plato: <nil>, Hertz: <nil>, Engine: parlia}"
INFO [07-15|02:32:38.074|eth/backend.go:190]           Initialising Ethereum protocol           network=56 dbversion=<nil>
INFO [07-15|02:32:38.075|consensus/parlia/parlia.go:685] Stored checkpoint snapshot to disk       number=0 hash=0d2184..d57b5b
INFO [07-15|02:32:39.140|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=500,004 hash=490dcd..376522 elapsed=1.065s
INFO [07-15|02:32:47.310|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=4,100,040 hash=3f4550..2533a5 elapsed=9.235s
INFO [07-15|02:32:56.765|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=6,400,063 hash=cdf101..a23bc9 elapsed=18.690s
INFO [07-15|02:33:05.235|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=8,400,083 hash=ca58bc..b715c0 elapsed=27.160s
INFO [07-15|02:33:13.261|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=10,500,104 hash=0d0b17..d9aa0c elapsed=35.186s
INFO [07-15|02:33:21.264|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=12,200,121 hash=470f67..025253 elapsed=43.189s
INFO [07-15|02:33:30.094|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=13,900,138 hash=12091f..e79dbd elapsed=52.019s
WARN [07-15|02:33:38.036|ethdb/leveldb/leveldb.go:380]    Database compacting, degraded performance database=<path redacted>/geth_data/geth/chaindata
INFO [07-15|02:33:38.459|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=15,500,154 hash=3640fa..1b2a24 elapsed=1m0.383s
INFO [07-15|02:33:48.276|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=17,300,172 hash=717e57..873c83 elapsed=1m10.201s
INFO [07-15|02:33:59.561|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=19,200,191 hash=1b8c07..f94ff2 elapsed=1m21.485s
INFO [07-15|02:34:08.604|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=20,300,202 hash=8f4f83..eee2a5 elapsed=1m30.528s
INFO [07-15|02:34:19.390|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=21,300,212 hash=8fee34..ce0d65 elapsed=1m41.314s
INFO [07-15|02:34:32.168|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=22,300,222 hash=c40342..acca60 elapsed=1m54.093s
WARN [07-15|02:34:38.048|ethdb/leveldb/leveldb.go:380]    Database compacting, degraded performance database=<path redacted>/geth_data/geth/chaindata
INFO [07-15|02:34:43.948|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=23,300,232 hash=635299..d3eb98 elapsed=2m5.873s
INFO [07-15|02:34:54.952|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=24,300,242 hash=f6e0db..6ddf17 elapsed=2m16.876s
INFO [07-15|02:35:05.447|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=25,400,253 hash=c4f3b1..5681cf elapsed=2m27.371s
INFO [07-15|02:35:20.940|core/rawdb/chain_iterator.go:74] Initializing database from freezer       total=26,856,810 number=26,500,264 hash=d380f5..9d1d9b elapsed=2m42.865s
INFO [07-15|02:35:21.623|core/rawdb/chain_iterator.go:85] Initialized database from freezer        blocks=26,856,810 elapsed=2m43.547s
INFO [07-15|02:35:22.305|consensus/parlia/parlia.go:685]  Stored checkpoint snapshot to disk       number=26,766,800 hash=92fe4c..2adbca
ERROR[07-15|02:35:22.522|consensus/parlia/snapshot.go:260] unauthorized validator         validators="map[0x0BAC492386862aD3dF4B666Bc096b0505BB694Da:0xc0014e7f40 0x2465176C461AfB316ebc773C61fAEe85A6515DAA:0xc0014e7fc0 0x295e26495CEF6F69dFA69911d9D8e4F3bBadB89B:0xc0014e7640 0x2D4C407BBe49438ED859fe965b140dcF1aaB71a9:0xc0014e79c0 0x35EBb5849518aFF370cA25E19e1072cC1a9FAbCa:0xc0014e7ac0 0x3f349bBaFEc1551819B8be1EfEA2fC46cA749aA1:0xc002180000 0x61Dd481A114A2E761c554B641742C973867899D3:0xc002180080 0x685B1ded8013785d6623CC18D214320b6Bb64759:0xc0014e7cc0 0x70F657164e5b75689b64B7fd1fA275F334f28e18:0xc0014e7bc0 0x72b61c6014342d914470eC7aC2975bE345796c2b:0xc0014e7d40 0x7AE2F5B9e386cd1B50A4550696D957cB4900f03a:0xc0014e7ec0 0x8b6C8fd93d6F4CeA42Bbb345DBc6F0DFdb5bEc73:0xc0014e7c40 0x9bB832254BAf4E8B4cc26bD2B52B31389B56E98B:0xc0014e7840 0x9F8cCdaFCc39F3c7D6EBf637c9151673CBc36b88:0xc0014e7dc0 0xb218C5D6aF1F979aC42BC68d98A5A0D796C6aB01:0xc0014e76c0 0xb4dd66D7c2C7E57F628210187192fb89d4b99dD4:0xc0014e78c0 0xBe807Dddb074639cD9fA61b47676c064fc50D62C:0xc0014e7940 0xD1d6bF74282782B0b3eb1413c901D6eCF02e8e28:0xc0014e7740 0xE9AE3261a475a27Bb1028f140bc2a7c843318afD:0xc0014e77c0 0xee226379dB83CfFC681495730c11fDDE79BA4c0C:0xc0014e7e40 0xEF0274E31810C9Df02F98FAFDe0f841F4E66a1Cd:0xc0014e75c0]"
ERROR[07-15|02:35:22.522|consensus/parlia/parlia.go:1734]  Unexpected error when getting snapshot   error="unauthorized validator: 0xa6f79B60359f141df90A0C745125B131cAAfFD12" blockNumber=26,856,809 blockHash=9a12a3..447d11
INFO [07-15|02:35:22.523|core/blockchain.go:671]           Loaded most recent local header          number=26,856,809 hash=9a12a3..447d11 td=53,374,987 age=3mo2w4d
INFO [07-15|02:35:22.523|core/blockchain.go:672]           Loaded most recent local full block      number=0          hash=0d2184..d57b5b td=1          age=3y3mo1w
INFO [07-15|02:35:22.523|core/blockchain.go:673]           Loaded most recent local fast block      number=26,856,809 hash=9a12a3..447d11 td=53,374,987 age=3mo2w4d
INFO [07-15|02:35:22.981|consensus/parlia/parlia.go:685]   Stored checkpoint snapshot to disk       number=26,766,800 hash=92fe4c..2adbca
```

The node will now repeatedly re-download the same blocks after the checkpoint block and fail to validate them.

Note that I added additional logging, including dumping the validator set when the unauthorized validator error occurs, so line numbers might not exactly match v1.2.8.

#### Steps to reproduce the behaviour

```
rm -rf geth_data/geth/chaindata
bsc/build/bin/geth --config config.toml --datadir geth_data --datadir.ancient <path redacted>/ancient --verbosity 3 --syncmode=full --cache 0 --log.debug --txlookuplimit 0 --http.corsdomain '*' init genesis.json
bsc/build/bin/geth --config config.toml --datadir geth_data --datadir.ancient <path redacted>/ancient --verbosity 3 --syncmode=full --cache 0 --log.debug --txlookuplimit 0 --http.corsdomain '*' console
```

#### Additional information

First time this occurred was shortly after I reported another issue #1345 and it was fixed in #1381 - may be related. I received this error shortly after #1381, then didn't touch the node for a while as it's not actually important for me.

**Comment by 0xLyon** [comment]
Hi there, can you kindly advise what your node startup commands are that is being used and also whether you are using a specific snapshot?

**Comment by immibis** [comment]
The commands are shown above. The freezer snapshot is whatever it was when I last played with BSC around March.

**Comment by 0xLyon** [comment]
Can you advise what version you were using back in March before you started the node again and upgraded to 1.2.8? Was there any versions that you have skipped by any chance?

**Comment by immibis** [comment]
Not sure, sorry. It would have been shortly after #1381 was fixed. Were there any bugs in Parlia consensus around that time? I think it's unlikely that a consensus bug would cause the network to make 90000 invalid blocks or whatever the limit is before they go into the freezer. If there was a consensus bug it should have only impacted the top few blocks and the node should've been stuck there. Plus BNB wouldn't be worth any money.
